pyccel/parser/semantic.py

```diff
@@ -2904,11 +2904,14 @@ def _visit_If(self, expr, **settings):
         variables = [v for branch in var_shapes for v in branch]
 
         for v in variables:
-            shape_branch1 = var_shapes[0][v]
-            if not all(v in branch_shapes.keys() for branch_shapes in var_shapes) \
-                    or not all(shape_branch1==branch_shapes[v] \
-                                for branch_shapes in var_shapes[1:]):
+            if v not in var_shapes[0]:
                 v.set_changeable_shape()
+            else:
+                shape_branch1 = var_shapes[0][v]
+                if not all(v in branch_shapes.keys() for branch_shapes in var_shapes) \
+                        or not all(shape_branch1==branch_shapes[v] \
+                                    for branch_shapes in var_shapes[1:]):
+                    v.set_changeable_shape()
```

Comment:
I think you can merge this if statement into one if.

Reply:
Can you suggest how in a way that avoids indexing `var_shapes` more times than necessary?

Reply:
```python
if v not in var_shapes[0] or not all(v in branch_shapes.keys() for branch_shapes in var_shapes) \
                        or not all(var_shapes[0][v]==branch_shapes[v] \
                                    for branch_shapes in var_shapes[1:]):
   v.set_changeable_shape()
```

Reply:
It was written as above to avoid doing:
```
all(var_shapes[0][v]==branch_shapes[v] \
                                    for branch_shapes in var_shapes[1:]):
```
which calls `var_shapes[0][v]` several times, when it only needs to be done once

Reply:
I don't know if python will optimize that, and read the value of var_shapes[0][v] once.

Reply:
XD normally "python" and "optimise" are not words that go together. I am happy to go back to this though, it may flip back later in the review though. It came from this discussion:
https://github.com/pyccel/pyccel/pull/921#discussion_r668750486

Reply:
In fact this:
```
if v not in var_shapes[0] or not all(v in branch_shapes.keys() for branch_shapes in var_shapes) \
                        or not all(var_shapes[0][v]==branch_shapes[v] \
                                    for branch_shapes in var_shapes[1:]):
   v.set_changeable_shape()
```
Can be written more simply as:
```
if not all(v in branch_shapes.keys() for branch_shapes in var_shapes) \
                        or not all(var_shapes[0][v]==branch_shapes[v] \
                                    for branch_shapes in var_shapes[1:]):
   v.set_changeable_shape()
```
So it seems the discussion above was the cause of this bug

Reply:
better: d0f141d0 ?